### PR TITLE
Reject

### PR DESCRIPTION
**Pull request by kickb71**
This Part Pull request solves the Feature part, Rejection Reasons part and Creation of DB table.

The issue is #802 
 
**Expected behavior**
- [x] Complete the feature of Rejection
- [x] Create  Database table
- [x] Use the Database table to store the values

Note: All the single quotes in the JS file changes to double quotes and the spaces because of the formatter

### CODE REVIEW

File: database/migrations/2021_09_06_081550_hr_application_rejection_reasons.php [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+     * @return void
+     */
+    public function up()
+    {
+        Schema::create('hr_application_rejection_reasons', function (Blueprint $table) {
+            $table->increments('id');
+            $table->integer('hr_application_round_id')->unsigned();
+            $table->foreign('hr_application_round_id')->references('id')->on('hr_application_round');
+            $table->string('reason_title', 255);
+            $table->text('reason_comment');
```
Comment (GauravGusain98) [inline database/migrations/2021_09_06_081550_hr_application_rejection_reasons.php:21]: I think this should be nullable. Just check if we can leave the comment field empty or not.

File: resources/js/app.js [hunk 2]
```diff
@@ -9,24 +9,24 @@ require("./bootstrap");
 import "jquery-ui/ui/widgets/datepicker.js";
```
Comment (GauravGusain98) [inline resources/js/app.js:9]: It is very difficult to find the changes you have made with these many formatting changes. So only format the code that you have added.

File: resources/views/hr/application/rejection-modal.blade.php [hunk 3]
```diff
@@ -41,6 +41,28 @@
                     <span aria-hidden="true">&times;</span>
                 </button>
             </div>
+                <div class="form-check">
+                    <form action="#" class="textbox">
+                        <h5><b> Select Reasons: </b></h5>
+                        <input type="checkbox" id="rejectReason1"> No responce <br>
```
Comment (GauravGusain98) [inline resources/views/hr/application/rejection-modal.blade.php:47]: Whenever you are using a text inside a blade tag, always sanitize it using `{{ __('') }}`

```suggestion
                        <input type="checkbox" id="rejectReason1"> {{ __('No response') }} <br>
```

File: resources/views/hr/application/rejection-modal.blade.php [hunk 4]
```diff
@@ -41,6 +41,28 @@
...
+                        <label for="Rejection">
+                            <input type="checkbox" id="rejectReason2" /> Skills mismatch
+                        </label>
+                        <br />
+                        <div id="txtReasonRejected1" style="display: none">
+                            <input type="text" id="Reasons1" />
+                        </div>
+                        <input type="checkbox" id="rejectReason3"> Culture mismatch <br>
+                        <input type="checkbox" id="rejectReason4"> Salary  expectation mismatch <br>
```
Comment (GauravGusain98) [inline resources/views/hr/application/rejection-modal.blade.php:56]:
```suggestion
                        <input type="checkbox" id="rejectReason4"> {{ __('Salary expectation mismatch') }} <br>
```

File: resources/views/hr/application/rejection-modal.blade.php [hunk 5]
```diff
@@ -41,6 +41,28 @@
...
+                        <label for="Rejection">
+                            <input type="checkbox" id="rejectReason5" /> Not enough knowledge/inclination for ColoredCow
+                        </label>
+                        <br />
+                        <div id="txtReasonRejected2" style="display: none">
```
Comment (GauravGusain98) [inline resources/views/hr/application/rejection-modal.blade.php:61]:
```suggestion
                        <div id="txtReasonRejected2" class="d-none">
```

File: resources/views/layouts/app.blade.php [hunk 6]
```diff
@@ -21,6 +21,9 @@
     <script type="text/javascript" src="{{ URL::asset('js/app.js') }}"></script>
     <script src="https://cdnjs.cloudflare.com/ajax/libs/Chart.js/2.8.0/Chart.js"></script>
 
+    <!-- JQuery -->
+    <script type="text/javascript" src="http://ajax.googleapis.com/ajax/libs/jquery/1.8.3/jquery.min.js"></script>
```
Comment (GauravGusain98) [inline resources/views/layouts/app.blade.php:25]: Why you have added this?

**Review by GauravGusain98 — CHANGES_REQUESTED** [review]
Please check the comments.

File: database/migrations/2021_09_06_081550_hr_application_rejection_reasons.php (hunk 1 above)
Comment (kickb71) [reply database/migrations/2021_09_06_081550_hr_application_rejection_reasons.php:21]: It will automatically add text when you click check box that's why its not nullable

File: resources/js/app.js (hunk 2 above)
Comment (kickb71) [reply resources/js/app.js:9]: I did only use the formater app it made changes every where

File: resources/views/hr/application/rejection-modal.blade.php (hunk 3 above)
Comment (kickb71) [reply resources/views/hr/application/rejection-modal.blade.php:47]: Sure let me change it  for all

File: resources/views/layouts/app.blade.php (hunk 6 above)
Comment (kickb71) [reply resources/views/layouts/app.blade.php:25]: > Why you have added this?

I have added this Jquery for when a check box is clicked a textbox should appear that's why I need jquery

File: resources/js/app.js (hunk 2 above)
Comment (GauravGusain98) [reply resources/js/app.js:9]: So when you format a document. There is option to format only changes. So just use that option. You can find it in settings.

File: resources/views/layouts/app.blade.php (hunk 6 above)
Comment (GauravGusain98) [reply resources/views/layouts/app.blade.php:25]: But we are using jquery earlier in app.js and it was working. Did you tried testing the code without importing this line. I would suggest testing it once without this line.

File: resources/js/app.js (hunk 2 above)
Comment (kickb71) [reply resources/js/app.js:9]: I will change it.  It won't happen again!

File: Modules/HR/Config/config.php [hunk 7]
```diff
@@ -127,4 +127,11 @@
         'max-reserved-allowed' => 3,
     ],
     'offer-letters-dir' => 'offer-letters',
+    'reasons-for-rejections' => [
+        'No response',
+        'Skills mismatch',
+        'Culture mismatch',
+        'Salary expectation mismatch',
+        'Not enough knowledge/inclination for ColoredCow'
```
Comment (P4NK4J) [inline Modules/HR/Config/config.php:135]: @kickb71 I think we should keep key=>value pair in the config, and store the slug in the database, since the labels may change in the future. Let me know your thoughts on it.

File: Modules/HR/Http/Requests/Recruitment/ApplicationRoundRequest.php [hunk 8]
```diff
@@ -51,6 +51,7 @@ public function rules()
                 $rules['mail_to_applicant.confirm.body'] = 'nullable|string|required_with:send_mail_to_applicant';
             }
         } elseif (request()->input('action') == 'reject') {
+            $rules['reject_reason'] = 'nullable|array';
```
Comment (P4NK4J) [inline Modules/HR/Http/Requests/Recruitment/ApplicationRoundRequest.php:54]: @kickb71 Any reason why nullable rule is being used here ?

File: Modules/HR/Routes/web.php [hunk 9]
```diff
@@ -65,7 +65,8 @@
             Route::post('/applicant', 'ApplicantController@store')->name('hr.applicant.store');
             Route::post('/excel-import', 'ApplicantController@importExcel')->name('hr.applications.excel-import');
 
-            Route::resource('applications/rounds', 'ApplicationRoundController')->only(['store', 'update']);
+            Route::resource('applications/rounds', 'ApplicationRoundController')->only(['store', 'update', 'storeReason']);
+            // Route::post('/applicationround/{applicationRound}/store', 'ApplicationRoundController@storeReason')->name('reasons.post');
```
Comment (P4NK4J) [inline Modules/HR/Routes/web.php:69]: @kickb71 Any specific reason, why the commented code is left in this file ?

File: database/migrations/2021_09_06_081550_hr_application_rejection_reasons.php [hunk 10]
```diff
@@ -0,0 +1,35 @@
+<?php
+
+use Illuminate\Database\Migrations\Migration;
+use Illuminate\Database\Schema\Blueprint;
+use Illuminate\Support\Facades\Schema;
+
+class HrApplicationRejectionReasons extends Migration
```
Comment (P4NK4J) [inline database/migrations/2021_09_06_081550_hr_application_rejection_reasons.php:7]: Why `Hr` naming convention is used here while at all other places `HR` is mentioned ? Let's use similar conventions everywhere.

File: resources/views/hr/application/rejection-modal.blade.php [hunk 11]
```diff
@@ -41,6 +41,18 @@
                     <span aria-hidden="true">&times;</span>
                 </button>
             </div>
+            <div class="form-check mt-3">
+                <p class="font-weight-bold">Select Reasons</p>
+                @foreach (config('hr.reasons-for-rejections') as $index => $reasonForRejection)
+                    <div class="rejection-reason-block">
+                        <label for="reasonTitle{{ $index }}">
+                            <input type="checkbox" class="reject-reason mr-1" id="reasonTitle{{ $index }}" name='reject_reason[{{ $index }}][title]' value='{{ $reasonForRejection }}'>{{ $reasonForRejection }}<br>
```
Comment (P4NK4J) [inline resources/views/hr/application/rejection-modal.blade.php:49]: I think we should use slug and value pair here, since the vocabulary for reasons may change in the future, we could have constant slug saved in the database then. Let me know your thoughts,

**Review by P4NK4J — CHANGES_REQUESTED** [review]
@kickb71 Left some comments, please take a look.

File: Modules/HR/Config/config.php (hunk 7 above)
Comment (kickb71) [reply Modules/HR/Config/config.php:135]: If we change the key value means it will be hard. In for loop in the blade file and javascript file I have to change in both places. If we use a new value means we should just add here the new value but if we use the key-value means we have to change all the files for future use. 
@P4NK4J  I think it should be the same. If it should be changed means I happy to change it

File: Modules/HR/Routes/web.php (hunk 9 above)
Comment (kickb71) [reply Modules/HR/Routes/web.php:69]: I forgot to remove it. Let me remove it

File: Modules/HR/Http/Requests/Recruitment/ApplicationRoundRequest.php (hunk 8 above)
Comment (kickb71) [reply Modules/HR/Http/Requests/Recruitment/ApplicationRoundRequest.php:54]: Yes. If you reject a person for only 2 reasons but all the others must not save that is why I gave nullable.

File: database/migrations/2021_09_06_081550_hr_application_rejection_reasons.php (hunk 10 above)
Comment (kickb71) [reply database/migrations/2021_09_06_081550_hr_application_rejection_reasons.php:7]: I can't understand what are trying to say?

File: resources/views/hr/application/rejection-modal.blade.php (hunk 11 above)
Comment (kickb71) [reply resources/views/hr/application/rejection-modal.blade.php:49]: I think it will hard to pass in this foreach

File: Modules/HR/Config/config.php (hunk 7 above)
Comment (P4NK4J) [reply Modules/HR/Config/config.php:135]: @kickb71 , I am saying the `key` will be saved in the database, which would never get changed. If there is any change to the reason let's say `No response` is changed to `No response given`, that would be easy to make only in the config, without changing it from the database, hence recommending to use the following
```suggestion
        'no-response' => 'No response',
        ...
```

File: Modules/HR/Http/Requests/Recruitment/ApplicationRoundRequest.php (hunk 8 above)
Comment (P4NK4J) [reply Modules/HR/Http/Requests/Recruitment/ApplicationRoundRequest.php:54]: If the others are not selected from the form field, they won't be saved then.

File: Modules/HR/Routes/web.php (hunk 9 above)
Comment (P4NK4J) [reply Modules/HR/Routes/web.php:69]: I can see it hasn't been removed. Let me know if there is something which I am missing.

File: database/migrations/2021_09_06_081550_hr_application_rejection_reasons.php (hunk 10 above)
Comment (P4NK4J) [reply database/migrations/2021_09_06_081550_hr_application_rejection_reasons.php:7]: Let me know if it's clear after today's discussion

File: resources/views/hr/application/rejection-modal.blade.php (hunk 11 above)
Comment (P4NK4J) [reply resources/views/hr/application/rejection-modal.blade.php:49]: Instead of $index, you use a slug in the value, and reason name in the label, while for the integer value you can initialize a different variable. Let me know if that makes sense.

**Review by P4NK4J — CHANGES_REQUESTED** [review]
@kickb71  Left some comments, please check.

File: Modules/HR/Config/config.php (hunk 7 above)
Comment (kickb71) [reply Modules/HR/Config/config.php:135]: The same thing happens without the key. I don't know the difference

File: Modules/HR/Http/Requests/Recruitment/ApplicationRoundRequest.php (hunk 8 above)
Comment (kickb71) [reply Modules/HR/Http/Requests/Recruitment/ApplicationRoundRequest.php:54]: There are two things the title and comment box we can leave the comment box empty that's why its nullable

**Comment by P4NK4J** [comment]
@kickb71 , Removing the status ready for review, I think there are some changes need to be made in this PR, as per our discussion.

File: database/migrations/2021_09_06_081550_hr_application_rejection_reasons.php (hunk 10 above)
Comment (P4NK4J) [reply database/migrations/2021_09_06_081550_hr_application_rejection_reasons.php:7]: @kickb71 Please resolve feedback before you assign for review

**Review by P4NK4J — COMMENTED** [review]
Left one comment, please check @kickb71

File: resources/views/hr/application/rejection-modal.blade.php (hunk 11 above)
Comment (P4NK4J) [reply resources/views/hr/application/rejection-modal.blade.php]: @kickb71 Also, you are not saving slugs in the value here, please make changes.

File: database/migrations/2021_09_06_081550_hr_application_rejection_reasons.php (hunk 10 above)
Comment (kickb71) [reply database/migrations/2021_09_06_081550_hr_application_rejection_reasons.php:7]: Changed!

File: resources/views/layouts/app.blade.php [hunk 12]
```diff
@@ -14,13 +14,17 @@
 
     <!-- Styles -->
     <link href="{{ mix('css/app.css') }}" rel="stylesheet">
-    <link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.7.0/css/font-awesome.min.css" rel="stylesheet">
+    {{-- <link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.7.0/css/font-awesome.min.css" rel="stylesheet"> --}}
```
Comment (P4NK4J) [inline resources/views/layouts/app.blade.php:17]: Why this CDN is commented @kickb71 ? Is it unused everywhere ?

**Review by P4NK4J — COMMENTED** [review]
Left a comment, let me know your thoughts

File: resources/views/layouts/app.blade.php (hunk 12 above)
Comment (kickb71) [reply resources/views/layouts/app.blade.php]: No i vas working on another issue that is why commented

**Review by GauravGusain98 — APPROVED** [review]
Looks good

**Review by P4NK4J — APPROVED** [review]
